Core/buffer serialization

RiccardoRossi opened the pull request:
this PR provides initial capabilities for a StreamSerializer which serializes data to a string instead of to a file.

it provides the interesting property that it can be pickled through python, meaning that for example it could be easily sent via internet (or via MPI)

RiccardoRossi committed:
adding serialization to string capabilities

RiccardoRossi committed:
adding a StreamSerializer wrapper

RiccardoRossi committed:
cleaning up the serialization test

RiccardoRossi committed:
cleanup

RiccardoRossi committed:
adding testing for model serialization and pickling2

philbucher commented:
cool feature

this works with any class that can be serialized with the Serializer right?

RiccardoRossi commented:
@philbucher correct! and this also mean that you can send things over the internet...so it opens the doors to the ☁️

loumalouomega commented:
> @philbucher correct! and this also mean that you can send things over the internet...so it opens the doors to the cloud

https://www.youtube.com/watch?v=iDbyYGrswtg

loumalouomega commented on kratos/includes/serializer.h:
```diff
@@ -187,6 +187,8 @@ class KRATOS_API(KRATOS_CORE) Serializer
     ///@}
     ///@name Operations
     ///@{
+    TraceType GetTraceType(){return mTrace;}
```
Some docu?

RiccardoRossi replied:
well... the doc of this would read "Get Trace Type" :-D

i would say that the naming is pretty descriptive here...

RiccardoRossi replied:
ok added

roigcarlo commented on kratos/python/add_serializer_to_python.cpp:
```diff
@@ -25,6 +25,39 @@
 {
+
+        //this object is needed because data may contain binary data so the treatment of strings 
+    //in python makes impossible to use them (reasons related to locale)
```
Looking at this twice, I think either this comment should go in the py::bytes cast or simply add this class in the pybind11 namespace (but not in the module) and return the py::byte array here

RiccardoRossi replied:
i changed the comment and moved to the python namespace

jcotela commented:
One question: as I remember from using it a couple of years ago, our serialization mechanism truncates doubles (I think to single precision?). Is this loss of precision mitigated somehow here? Do you think that this could be a problem if we use this mechanism for simulation data? (as opposed to boundary conditions, which tend to have "nice" numbers)

RiccardoRossi commented:
@jcotela, i am doing absolutely nothing here about precision. I was assuming that the Serializer handles this correctly, if it is not so (and i simply ignore the answer to this) then we need to fix the "standard" serializer. In any case i would say your remark is not related to this PR.

Indeed the only difference here is that the StreamSerializer object writes to a stream of data instead of to a file. 
Actually the StreamSerializer object is purely a tool to provide a "nice" (read, a bit more pythonic)  interface, and under the hood it simply uses the standard Serializer object. 
Indeed we can discuss if this is the correct mechanism or if we should use a different one. Consider this as a first attempt...

roigcarlo commented:
In the topic of the double precision, I think the problem was while using the ASCII mode ( with the TRACE_ALL flag activated). I personally have used the serializer with doubles without any problem.

jcotela commented:
@roigcarlo it could well be that this was related to printing in ASCII mode (I think it was the only mode available in the very first version of the serializer?), I do not remember the details anymore. I guess that this would not be a problem, then. @RiccardoRossi I know that this PR is not directly modifying the serializer, I just wanted to clarify the issue before we adopted this as a definitive solution.

RiccardoRossi committed:
improving comments

RiccardoRossi commented:
@jcotela no worries. As i am telling also i am not sure if how i handled this is the "definitive" answer...

RiccardoRossi committed:
adding a comment about GetTraceType

pooyan-dadvand commented on kratos/python/add_serializer_to_python.cpp:
```diff
@@ -25,8 +25,47 @@
 {
+
+    //This is a Helper object to provide a simpler interface for serialization to a string instead of to a file
+    class StreamSerializer : public Serializer 
```
I would suggest to add this class in Internals namespace in order to show that it is not for external use....

RiccardoRossi replied:
Note that it is not "Internal" it is of the python interface.
I moved it to the Python namespace exactly for this reason.

pooyan-dadvand commented on kratos/python/add_serializer_to_python.cpp:
```diff
@@ -62,6 +105,26 @@ void  AddSerializerToPython(pybind11::module& m)
+    .def("__getstate__", [](StreamSerializer &self) { //METHOD NEEDED FOR PICKLE
+        /* Return a tuple that fully encodes the state of the object */
+        //note that we return a "bytes" object to avoid any "locale" conversion
+        return py::make_tuple(py::bytes(self.GetStringRepresentation()),self.GetTraceType());
```
The order shouldn't be inverse?

RiccardoRossi replied:
well, i am putting this in a tuple, and it is consistent with how i use it in line 125 (which is mandated by the constructor of StreamSerializer)

if we want to change the order i don't have any particular objection

pooyan-dadvand reviewed (commented):
@RiccardoRossi After talking with you and think more about it I arrive to the point that it would be more clear if we define the `operator<<` for serializer to dump its buffer. Then we don't need this class in the middle and the interface would be clear and reusable for other usages in C++.

What is your opinion?

RiccardoRossi commented:
@pooyan-dadvand this class "in the middle" is not needed for the interface, but only for making the construction nicer.

The problem is that the standard serializer is constructed by 
 
       Serializer(filename, tracetype) --> output to a file
       Serializer(tracetype) --> output to a buffer

No doubt i could define a function "SetBufferData" so that

       Serializer(tracetype).SetBufferData(data)

would work correctly. The point is how to prevent someone from doing

      Serializer(filename, tracetype).SetBufferData(data)

which makes no sense... 
The class i created is an attempt of preventing users from possible wrong usages, so that one defines
  
      StreamSerializer(data)

and gets a behaviour that is completely equivalent to doing

      Serializer(tracetype).pGetBuffer() << data

(a syntax that is not possible in python)

philbucher commented:
I suggest taking a look at the codacy suggestions

pooyan-dadvand commented:
I see your point of view that we are changing the complete behaviour of serializer object passing different objects. So maybe a solution would be to apply this not only in Python but also in cpp. 

IMO this can be done by driving two classes from Serializer knowing that non of the save and load methods can be override but one can define the stream buffer. This may also open the door for a dedicated MPI serializer.

RiccardoRossi committed:
using explicitly FileSerializer instead of Serializer to avoid a warning

RiccardoRossi committed:
using StreamSerializer insteadof Serializer

RiccardoRossi committed:
removing commodity constructor to leave just one which obliges to define the buffer explicitly

RiccardoRossi committed:
improving pickle interface to avoid a deprecation warning and deprecating Serializer constructors

RiccardoRossi committed:
modernizing to use FileSerializer instead of Serializer

RiccardoRossi committed:
using StreamSerializer

RiccardoRossi committed:
addressing codacy comments

RiccardoRossi committed:
adding simple derived classes to simplify usage

RiccardoRossi committed:
minor

RiccardoRossi committed:
making const the GetTraceType

RiccardoRossi committed:
cleanup

RiccardoRossi commented:
Ok, so with the latest changes i implemented the proposal we discussed with Pooyan.

In essence i removed the commodity constructors from the Serializer class, leaving just one defined as

        Serializer(Buffer, TraceType)

This is quite unconforable to use since one needs to properly pass the buffer in the constructor.

then i added two derived classes named: 

- StreamSerializer --> constructs     Serializer(basic_iostream, TraceType )
- FileSerializer       --> constructs:    Serializer(filestream, TraceType)

aside of this i fixed codacy warnings and polished the pickle interface to get rid of warnings

pooyan-dadvand commented on kratos/includes/file_serializer.h:
```diff
@@ -0,0 +1,64 @@
+
+namespace Kratos
+{
+    //This is a Helper object to provide a simpler interface for serialization to a file
```
//This class provides a simpler interface for serialization to a file
// Note that you may not override any load or save method of the Serializer. They are not virtual.

pooyan-dadvand commented on kratos/includes/stream_serializer.h:
```diff
@@ -0,0 +1,67 @@
+#include "includes/define.h"
+#include "includes/serializer.h"
+
+//This is a Helper object to provide a simpler interface for serialization to a string instead of to a file
```
//This class provides a simpler interface for serialization to a string instead of to a file
// Note that you may not override any load or save method of the Serializer. They are not virtual.

pooyan-dadvand reviewed (dismissed):
The are two small suggestions to improve the comments if you like to apply them.

RiccardoRossi committed:
applying pooyan's correction on the comment

pooyan-dadvand commented:
Thanks @RiccardoRossi but they were two comment! :stuck_out_tongue_winking_eye:

RiccardoRossi committed:
adding pooyan's comment

pooyan-dadvand reviewed (dismissed):
Thanks!

philbucher committed:
Merge branch 'master' into core/buffer_serialization

RiccardoRossi committed:
merging master into the branch

RiccardoRossi committed:
small modification to test both FileSerializer and StreamSerializer

pooyan-dadvand reviewed (dismissed):
👍

RiccardoRossi commented:
@roigcarlo any clue of why AppVeyor failed on this? there is no error in the compilation...

philbucher commented:
I think it is timing out
EDIT yep also my build just timed out ...

For some reason today they need much longer (> 1hr) than 2 days ago (~45min)

RiccardoRossi commented:
now it compiled again in 45min, meaning that it was something with their systems...

RiccardoRossi committed:
merging master

pooyan-dadvand reviewed (approved):
👍
